src/main/ruby/core/regexp.rb

```diff
@@ -381,3 +406,40 @@ def to_s
     self[0]
   end
 end
+
+Truffle::KernelOperations.define_hooked_variable(
+  :$~,
+  Truffle::Graal.always_split( -> b { Truffle::RegexpOperations.last_match(b) }),
+  Truffle::Graal.always_split( -> v, b { Truffle::RegexpOperations.set_last_match(v, b) }))
+
+Truffle::KernelOperations.define_hooked_variable_with_is_defined(
+  :$`,
+  Truffle::Graal.always_split(
+    -> b { match = Truffle::RegexpOperations.last_match(b)
+           match.pre_match if match }),
+  -> { raise SyntaxError, "Can't set variable $`"},
+  Truffle::Graal.always_split( -> b { 'global-variable' if Truffle::RegexpOperations.last_match(b) }))
+
+Truffle::KernelOperations.define_hooked_variable_with_is_defined(
+  :$',
+  Truffle::Graal.always_split(
+    -> b { match = Truffle::RegexpOperations.last_match(b)
+           match.post_match if match }),
+  -> { raise SyntaxError, "Can't set variable $'"},
+  Truffle::Graal.always_split( -> b { 'global-variable' if Truffle::RegexpOperations.last_match(b) }))
+
+Truffle::KernelOperations.define_hooked_variable_with_is_defined(
+  :$&,
+  Truffle::Graal.always_split(
+    -> b { match = Truffle::RegexpOperations.last_match(b)
+           match[0] if match }),
+  -> { raise SyntaxError, "Can't set variable $&"},
+  Truffle::Graal.always_split( -> b { 'global-variable' if Truffle::RegexpOperations.last_match(b) }))
+
+Truffle::KernelOperations.define_hooked_variable_with_is_defined(
+  :$+,
+  Truffle::Graal.always_split(
+    -> b { match = Truffle::RegexpOperations.last_match(b)
+           match.captures.reject { |m| m.nil? }.last if match }),
+  -> { raise SyntaxError, "Can't set variable $+"},
+  Truffle::Graal.always_split( -> b { 'global-variable' if Truffle::RegexpOperations.last_match(b) }))
```

Comment:
The splitting is very verbose.
Maybe always split in `define_hooked_variable_with_is_defined`?
Or at least when it takes a binding argument (arity 1)?
Otherwise you could also define a helper in a local var like
```ruby
split = -> proc { Truffle::Graal.always_split(proc) }
split.(-> b { ... })
```